app-ios/tutanota/Sources/Contacts/ContactsSynchronization.swift

```diff
@@ -0,0 +1,336 @@
+
+  storeContactsNotFound = Array(nativeToTutaContactIdentifiers.values)
+  return contacts
+}
+
+private func saveNativeContacts(_ contacts: TutaToNativeContacts, forUserId userId: String, deleteOldContacts delete: Bool) throws {
+  let store = CNContactStore()
+  let saveRequest = CNSaveRequest()
+  let contactGroup = try getTutaContactGroup()
+  for contact in contacts {
+    contact.value.applyToSaveRequest(request: saveRequest, tutaContactGroup: contactGroup)
+  }
+  try store.execute(saveRequest)
+  
+  writeSavedLocalContactsIdentifierMappings(contacts.mapValues { contact in contact.contact.identifier }, forUserId: userId, merge: !delete)
+}
+
+/// Gets the Tuta contact group, creating it if it does not exist.
+private func getTutaContactGroup() throws -> CNGroup {
+  let store = CNContactStore()
+  if let groupIdentifier = getTutaContactsGroupIdentifier() {
+    let result = try store.groups(matching: CNGroup.predicateForGroups(withIdentifiers: [groupIdentifier]))
+    if !result.isEmpty {
+      return result[0]
+    }
+    TUTSLog("can't get tuta contact group \(groupIdentifier) from native: likely deleted by user")
+  }
+  
+  let newGroup = CNMutableGroup()
+  newGroup.name = "Tuta"
+  
+  let saveRequest = CNSaveRequest()
+  saveRequest.add(newGroup, toContainerWithIdentifier: nil)
+  
+  try store.execute(saveRequest)
+  setTutaContactsGroupIdentifier(newGroup.identifier)
+  
+  return newGroup
+}
+
+private func removeNativeContacts(_ contacts: TutaToNativeContactIdentifiers, forUserId userId: String) throws {
+  var identifiersToRemove = Set(contacts.values)
+  
+  // we now need to create a request to remove all contacts from the user that match an id in idsToRemove
+  // it is OK if we are missing some contacts, as they are likely already deleted
+  let store = CNContactStore()
+  let fetch = CNContactFetchRequest(keysToFetch: [CNContactIdentifierKey] as [CNKeyDescriptor])
+  let save = CNSaveRequest()
+  var canDelete = false
+  try store.enumerateContacts(with: fetch) { contact, stopPointer in
+    guard identifiersToRemove.remove(contact.identifier) != nil else { return }
+    save.delete(contact.mutableCopy() as! CNMutableContact)
+    canDelete = true
+    
+    // stop early; nothing left to remove
+    if identifiersToRemove.isEmpty {
+      stopPointer.initialize(to: true)
+    }
+  }
+  
+  // nothing to delete?
+  if !canDelete {
+    return
+  }
+  
+  try store.execute(save)
+  
+  var oldIdentifiers = getSavedLocalContactsIdentifierMappings(forUserId: userId)
+  for contact in contacts {
+    oldIdentifiers[contact.key] = nil
+  }
+  writeSavedLocalContactsIdentifierMappings(oldIdentifiers, forUserId: userId, merge: false)
+}
+
+private func getContactsMappingsDictionary() -> [String:Any] {
+  return UserDefaults.standard.dictionary(forKey: CONTACTS_MAPPINGS) ?? [:]
+}
+
+private func writeContactsMappingsDictionary(_ dictionary: [String:Any]) {
+  UserDefaults.standard.set(dictionary, forKey: CONTACTS_MAPPINGS)
```

Comment:
this stuff should be in the class and userDefaults should be injected

Reply:
I agree, but I did not see that we injected UserDefaults anywhere else, so I chose not to do it. Would it make sense to write a user defaults provider of some kind?